### PR TITLE
move_group.cpp: seg fault bug fix

### PR DESCRIPTION
**Pull request by gaspers**
### Description
Fixed a bug with casting CallbackQueueInterface * to ros::CallbackQueue * without a check.
Without this additional check, creating an instance of MoveGroup in a nodelet causes a segmentation fault, because of different CallbackQueue implementations (nodelet::detail::CallbackQueue).

https://github.com/ros-planning/moveit/issues/425

### CODE REVIEW

**Comment by v4hn** [comment]
Thank you for the patch! So standard ROS indeed provides another type of callback structure...
I'm sorry, nobody thought of nodelets back then..

I don't really like to rely on RTTI to resolve this, but imo it's roscpp's fault for not exposing a processing-method for the queue via the abstract interface... So I don't see a cleaner way to keep all use-cases working.

I just checked and RTTI is required in *core* MoveIt already, but this patch will force user-code, i.e. everything using the MoveGroupInterface, to compile with it too. Granted, it's not something you switch-off everyday though..

I approve the patch. This is a plain bugfix and easy enough to understand.
Should be forward-ported to j/k.

**Comment by de-vri-es** [comment]
I'm not familiar with nodelets callback queue / threading model, but don't we run the risk of processing the wrong callback queue using `ros::spinOnce()`? If so that could potentially lead to race conditions if handlers are fired that the code doesn't except.

Another option may to to explicitly give a callback queue to the action client on construction and process that queue exactly when desired (of which we will know the exact type, since we created it, so also no RTTI required).

Alternatively, if there is no risk of race conditions, is there a reason not to go for `ros::spinOnce()` in all cases?

**Comment by gaspers** [comment]
ros::spinOnce() was the default before. That was changed by https://github.com/ros-planning/moveit_ros/pull/701 to support specific callbacks, with the problem of dropping nodelet support. That is the reason for not doing ros::spinOnce() in all cases.

**Comment by de-vri-es** [comment]
Right, but then by reverting to `ros::spinOnce()` we don't support specific callback queues any more if the callback queue is not a `ros::CallbackQueue`. Of course it's already broken for nodelets as it is, but we should be sure not to introduce possible race conditions.

**Comment by v4hn** [comment]
On Thu, Jan 26, 2017 at 10:07:50AM -0800, Maarten de Vries wrote:
> Right, but then by reverting to `ros::spinOnce()` we don't support specific callback queues any more
> if the callback queue is not a `ros::CallbackQueue`.

This explicitly was never the case. The requestor of the pull-request for specific callback queues was fine with that restriction.
This patch does not revert anything. It basically adds error handling to the previous request by replacing the static cast with
an RTTI-based dynamic one.

> Of course it's already broken for nodelets as it is, but we should be sure not to introduce possible race conditions.

The patch fixes the recently broken non-`ros::CallbackQueue` support by reverting to the previous behavior in that case.
So this shouldn't break anything that hasn't been broken for years..

**Comment by de-vri-es** [comment]
> The patch fixes the recently broken non-`ros::CallbackQueue` support by reverting to the previous behavior in that case. So this shouldn't break anything that hasn't been broken for years..

Nodelets weren't considered back then. Since this is specifically geared for nodelets it makes sense to look at the implications of that.

However, as far as I can tell from the source, the nodelet manager doesn't even use the default callback queue. It seems to create its own thread pool and manage everything from there. That means the `spinOnce()` may be a nop in the case of nodelets, but that's not a race condition. And since the nodelets are processing their callback queues in a separate thread pool I don't think it matters that the `spinOnce()` may not be doing anything.

More importantly, since all the `waitForAction` calls are in the constructor it should be safe anyway. Any code that is accessing the object before the constructor finished is wrong. So not our problem.

In other words: I think I'm convinced there's no danger of race conditions.

**Comment by gaspers** [comment]
> However, as far as I can tell from the source, the nodelet manager doesn't even use the default callback queue. It seems to create its own thread pool and manage everything from there. That means the spinOnce() may be a nop in the case of nodelets, but that's not a race condition. And since the nodelets are processing their callback queues in a separate thread pool I don't think it matters that the spinOnce() may not be doing anything.

Yes, you are right. Nodelets don't need ros::spinOnce() as the nodelet manager handles threads and processes message queues. Additionally, spinOnce() wouldn't help with different CallbackQueue implementations either, I think. SpinOnce() only calls the global CallbackQueue. So perhaps, I should remove the else-block? That would reduce the fix to preventing the broken cast.
In the case of nodelets, we tested the code without spinOnce successfully. In the default case of a global ros::CallbackQueue, the code should work exactly like spinOnce(). That was stated and tested in https://github.com/ros-planning/moveit_ros/pull/701.

**Comment by de-vri-es** [comment]
> Additionally, spinOnce() wouldn't help with different CallbackQueue implementations either
I think you're right. Maybe we should log a warning instead? Though it is possible that some other thread is already processing the relevant callback queue (like the nodelet queue), in which case the warning would be a bit of a nuisance.

**Comment by gaspers** [comment]
I integrated your suggestion, added a comment (for else branch) and tried to find a helpful warning text that is emitted once instead of calling spinOnce().

**Comment by v4hn** [comment]
Merged and cherry-picked to jade/kinetic. Thank you lots for creating the issue and the patch @gaspers !

**Comment by gaspers** [comment]
Thank you for the fast responses and for including our patch.
